File: app/src/main/java/br/com/vicentec12/mygames/ui/activity/AddGameActivity.java
package br.com.vicentec12.mygames.ui.activity;

import android.os.Bundle;
import android.view.View;

import androidx.appcompat.app.AppCompatActivity;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.snackbar.BaseTransientBottomBar;
import com.google.android.material.snackbar.Snackbar;
import com.google.android.material.textfield.TextInputLayout;

import br.com.vicentec12.mygames.R;
import br.com.vicentec12.mygames.data.model.Game;
import br.com.vicentec12.mygames.data.source.Callbacks;
import br.com.vicentec12.mygames.data.source.repository.GameRepository;
import br.com.vicentec12.mygames.util.InstantiateUtil;
import br.com.vicentec12.mygames.util.ValidationUtil;

public class AddGameActivity extends AppCompatActivity {

    private TextInputLayout _tilName;
    private TextInputLayout _tilYear;
    private FloatingActionButton _fabAddGame;

    private boolean isUpdate;
    private Game mSelectedGame;
    private GameRepository mGameRepository;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_add_game);
        init();
    }

    private void init() {
        _tilName = findViewById(R.id.til_add_game_name);
        _tilYear = findViewById(R.id.til_add_game_year);
        _fabAddGame = findViewById(R.id.fab_add_game);
        config();
    }

    private void config() {
        configUpdate();
    }

    private void configUpdate() {
        mSelectedGame = (Game) getIntent().getSerializableExtra("game");
        if (mSelectedGame != null) {
            isUpdate = true;
            _tilName.getEditText().setText(mSelectedGame.getName());
            _tilYear.getEditText().setText(mSelectedGame.getYear());
            _fabAddGame.setImageResource(R.drawable.ic_edit);
        }
    }

    private boolean validateFields() {
        removeFieldErrors();
        if (!ValidationUtil.validateEmptyField(this, _tilName))
            return false;
        if (!ValidationUtil.validateEmptyField(this, _tilYear))
            return false;
        return true;
    }

    private void removeFieldErrors() {
        ValidationUtil.removeErrorTextInputLayout(_tilName);
        ValidationUtil.removeErrorTextInputLayout(_tilYear);
    }

    public void fabEvent(View view) {
        if (validateFields()) {
            String name = _tilName.getEditText().getText().toString();
            String year = _tilYear.getEditText().getText().toString();
            mGameRepository = InstantiateUtil.instantialeGameRepository(this);
            if (isUpdate) {
                mSelectedGame.setName(name);
                mSelectedGame.setYear(year);
                updateGame(mSelectedGame);
            } else {
                Game game = new Game(name, year);
                insertGame(game);
            }
        }
    }

    private void insertGame(Game game) {
        GameRepository mGameRepository = InstantiateUtil.instantialeGameRepository(this);
        mGameRepository.insert(this, game, new Callbacks.OnLocalCallback() {
            @Override
            public void onSuccess(String message) {
                setResult(RESULT_OK);
                _tilName.getEditText().setText("");
                _tilYear.getEditText().setText("");
                Snackbar.make(_tilName, message, BaseTransientBottomBar.LENGTH_LONG)
                        .setAnimationMode(BaseTransientBottomBar.ANIMATION_MODE_SLIDE).show();
            }

            @Override
            public void onFailure(String message) {
                Snackbar.make(_tilName, message, BaseTransientBottomBar.LENGTH_LONG)
                        .setAnimationMode(BaseTransientBottomBar.ANIMATION_MODE_SLIDE).show();
            }
        });
    }

    private void updateGame(Game game) {
        mGameRepository.update(this, game, new Callbacks.OnLocalCallback() {
            @Override
            public void onSuccess(String message) {
                setResult(RESULT_OK);
                Snackbar.make(_tilName, message, BaseTransientBottomBar.LENGTH_LONG)
                        .setAnimationMode(BaseTransientBottomBar.ANIMATION_MODE_SLIDE).show();
            }

            @Override
            public void onFailure(String message) {
                Snackbar.make(_tilName, message, BaseTransientBottomBar.LENGTH_LONG)
                        .setAnimationMode(BaseTransientBottomBar.ANIMATION_MODE_SLIDE).show();
            }
        });
    }

}

File: app/src/main/java/br/com/vicentec12/mygames/presentation/add_game/AddGameActivity.kt
package br.com.vicentec12.mygames.presentation.add_game

import android.content.Context
import android.content.Intent
import android.os.Bundle
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import br.com.vicentec12.mygames.R
import br.com.vicentec12.mygames.databinding.ActivityAddGameBinding
import br.com.vicentec12.mygames.domain.model.Game
import br.com.vicentec12.mygames.extensions.*
import com.google.android.material.snackbar.BaseTransientBottomBar
import com.google.android.material.snackbar.Snackbar
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class AddGameActivity : AppCompatActivity() {

    private val mViewModel: AddGameViewModel by viewModels()

    private val mBinding by viewBinding(ActivityAddGameBinding::inflate)

    private val mSelectedGame: Game by lazy { intent.getParcelableExtra(EXTRA_GAME) ?: Game() }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(mBinding.root)
        initView()
    }

    private fun initView() {
        initBinding()
        initToolbar()
        initObservers()
        initListeners()
    }

    private fun initBinding() {
        mBinding.viewModel = mViewModel
        mBinding.lifecycleOwner = this@AddGameActivity
    }

    private fun initToolbar() {
        with(mBinding.lytToolbar.toolbar) {
            setTitle(if (mSelectedGame.id == 0L) R.string.title_activity_add_games else R.string.title_activity_edit_games)
            setSupportActionBar(this)
        }
    }

    private fun initObservers() {
        with(mViewModel) {
            message.observe(this@AddGameActivity) { mMessageEvent ->
                mMessageEvent.contentIfNotHandled?.let { mMessageId ->
                    Snackbar.make(
                        mBinding.tilAddGameName,
                        mMessageId,
                        BaseTransientBottomBar.LENGTH_LONG
                    )
                        .setAnimationMode(BaseTransientBottomBar.ANIMATION_MODE_SLIDE).show()
                }
            }
            success.observe(this@AddGameActivity) { mDataBaseEvent ->
                mDataBaseEvent.contentIfNotHandled?.let {
                    setResult(RESULT_OK)
                    mBinding.tilAddGameName.requestFocus()
                }
            }
            listConsoles(mSelectedGame)
        }
    }

    private fun initListeners() {
        mBinding.fabAddGame.setOnClickListener {
            if (validateFields())
                mViewModel.databaseEvent()
        }
    }

    private fun validateFields(): Boolean {
        with(mBinding) {
            removeFieldErrors()
            return when {
                !tilAddGameName.validateEmptyField() -> false
                !tilAddGameYear.validateEmptyField() -> false
                !tilAddGameYear.validateQtdCharacters(YEAR_QUANTITY) -> false
                !spnAddGameConsole.validateSelection(getString(R.string.error_message_spinner_selection)) -> false
                else -> true
            }
        }
    }

    private fun removeFieldErrors() {
        mBinding.tilAddGameName.removeError()
        mBinding.tilAddGameYear.removeError()
    }

    companion object {

        private const val EXTRA_GAME = "extra_game"

        fun newIntentInstance(mContext: Context, mGame: Game) =
            Intent(mContext, AddGameActivity::class.java).apply {
                putExtra(EXTRA_GAME, mGame)
            }

    }

}